File: avalon-modules/avalon-dataset/avalon-dataset-repository/src/test/java/io/github/lishangbu/avalon/dataset/repository/MoveCategoryRepositoryTest.java
package io.github.lishangbu.avalon.dataset.repository;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.github.lishangbu.avalon.dataset.configuration.DatasetRepositoryTestEnvironmentConfiguration;
import io.github.lishangbu.avalon.dataset.entity.MoveCategory;
import java.util.Optional;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.data.jdbc.DataJdbcTest;
import org.springframework.test.context.ContextConfiguration;

/**
 * @author lishangbu
 * @since 2025/4/15
 */
@ContextConfiguration(classes = DatasetRepositoryTestEnvironmentConfiguration.class)
@DataJdbcTest
@SuppressWarnings("SpringJavaInjectionPointsAutowiringInspection")
class MoveCategoryRepositoryTest {
  @Autowired private MoveCategoryRepository moveCategoryRepository;

  @Test
  void testFindByCategory() {
    Optional<MoveCategory> moveCategoryOptional =
        moveCategoryRepository.findByInternalName("damage");
    assertTrue(moveCategoryOptional.isPresent());
    assertEquals("damage", moveCategoryOptional.get().getName());
    assertEquals("Inflicts damage", moveCategoryOptional.get().getDescription());
  }

  @Test
  void testFindById() {
    Optional<MoveCategory> moveCategoryOptional = moveCategoryRepository.findById(0L);
    assertTrue(moveCategoryOptional.isPresent());
    assertEquals("damage", moveCategoryOptional.get().getName());
    assertEquals("Inflicts damage", moveCategoryOptional.get().getDescription());
  }
}

File: avalon-modules/avalon-dataset/avalon-dataset-repository/src/test/kotlin/io/github/lishangbu/avalon/dataset/repository/MoveCategoryRepositoryTest.kt
package io.github.lishangbu.avalon.dataset.repository

import io.github.lishangbu.avalon.dataset.entity.MoveCategory
import io.github.lishangbu.avalon.dataset.entity.dto.MoveCategorySpecification
import jakarta.annotation.Resource
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class MoveCategoryRepositoryTest : AbstractRepositoryTest() {
    @Resource
    private lateinit var moveCategoryRepository: MoveCategoryRepository

    @Test
    fun shouldInsertMoveCategorySuccessfully() {
        val moveCategory =
            MoveCategory {
                internalName = "unit-test-move-category"
                name = "单元测试类别"
                description = "单元测试描述"
            }

        val saved = moveCategoryRepository.save(moveCategory, SaveMode.INSERT_ONLY)

        assertTrue(saved.id > 0)
    }

    @Test
    fun shouldFindMoveCategoryById() {
        val moveCategory = requireNotNull(moveCategoryRepository.loadViewById(0L))

        assertEquals("0", moveCategory.id)
        assertEquals("damage", moveCategory.internalName)
        assertEquals("damage", moveCategory.name)
    }

    @Test
    fun shouldUpdateMoveCategoryById() {
        val moveCategory =
            moveCategoryRepository.save(
                MoveCategory {
                    internalName = "move-category-update"
                    name = "原始类别"
                    description = "原始描述"
                },
                SaveMode.INSERT_ONLY,
            )
        val id = moveCategory.id

        moveCategoryRepository.save(
            MoveCategory(moveCategory) {
                name = "更新后的类别"
            },
            SaveMode.UPSERT,
        )

        val updatedMoveCategory = requireNotNull(moveCategoryRepository.findNullable(id))
        assertEquals("更新后的类别", updatedMoveCategory.name)
    }

    @Test
    fun shouldDeleteMoveCategoryById() {
        val moveCategory =
            moveCategoryRepository.save(
                MoveCategory {
                    internalName = "move-category-delete"
                    name = "待删除类别"
                    description = "待删除描述"
                },
                SaveMode.INSERT_ONLY,
            )
        val deleteRecordId = moveCategory.id

        assertNotNull(moveCategoryRepository.findNullable(deleteRecordId))
        moveCategoryRepository.deleteById(deleteRecordId)
        assertNull(moveCategoryRepository.findNullable(deleteRecordId))
    }

    @Test
    fun shouldSelectListWithDynamicCondition() {
        val condition = MoveCategorySpecification(internalName = "damage")

        val results = moveCategoryRepository.listViews(condition)

        assertTrue(results.isNotEmpty())
        assertTrue(results.any { it.name == "damage" })
    }

    @Test
    fun shouldReturnAllMoveCategoriesWhenNoCondition() {
        val results = moveCategoryRepository.listViews(null)

        assertTrue(results.isNotEmpty())
        assertTrue(results.any { it.internalName == "damage" })
    }
}
